Adjust the loading order of preload

MisterRaindrop opened the pull request:
Prioritize loading the process_shared_preload_libraries.h
header file, and then load the shared_preload_libraries from postgresql.conf.

<!-- Thank you for your contribution to Apache Cloudberry (Incubating)! -->

Fixes #ISSUE_Number

### What does this PR do?
<!-- Brief overview of the changes, including any major features or fixes -->

### Type of Change
- [ ] Bug fix (non-breaking change)
- [ ] New feature (non-breaking change)
- [ ] Breaking change (fix or feature with breaking changes)
- [ ] Documentation update

### Breaking Changes
<!-- Remove if not applicable. If yes, explain impact and migration path -->

### Test Plan
<!-- How did you test these changes? -->
- [ ] Unit tests added/updated
- [ ] Integration tests added/updated
- [ ] Passed `make installcheck`
- [ ] Passed `make -C src/test installcheck-cbdb-parallel`

### Impact
<!-- Remove sections that don't apply -->
**Performance:**
<!-- Any performance implications? -->

**User-facing changes:**
<!-- Any changes visible to users? -->

**Dependencies:**
<!-- New dependencies or version changes? -->

### Checklist
- [ ] Followed [contribution guide](https://cloudberry.apache.org/contribute/code)
- [ ] Added/updated documentation
- [ ] Reviewed code for security implications
- [ ] Requested review from [cloudberry committers](https://github.com/orgs/apache/teams/cloudberry-committers)

### Additional Context
<!-- Any other information that would help reviewers? Remove if none -->

### CI Skip Instructions
<!--
To skip CI builds, add the appropriate CI skip identifier to your PR title.
The identifier must:
- Be in square brackets []
- Include the word "ci" and either "skip" or "no"
- Only use for documentation-only changes or when absolutely necessary
-->

---
<!-- Join our community:
- Mailing list: [dev@cloudberry.apache.org](https://lists.apache.org/list.html?dev@cloudberry.apache.org) (subscribe: dev-subscribe@cloudberry.apache.org)
- Discussions: https://github.com/apache/cloudberry/discussions -->

my-ship-it commented:
Hi @MisterRaindrop, thanks for the PR.
Why we change the order here. Could you please share the context.

MisterRaindrop commented:
@my-ship-it 
Normally, we load libraries (libs) in two places: one is in the header files, and the other is in the postgresql.conf file. In the past, our approach was to append the libraries that needed to be loaded from the header files to the shared_preload_libraries parameter in postgresql.conf. Since the libraries loaded from the header files are compiled into the kernel, we want them to have the highest priority. If some libraries need to be loaded immediately, it could cause conflicts. Therefore, we place the loading of libraries from the header files first, ensuring they are loaded during the compilation process before the libraries specified in the postgresql.conf configuration are loaded.

my-ship-it commented:
> @my-ship-it OK Normally, we load libraries (libs) in two places: one is in the header files, and the other is in the postgresql.conf file. In the past, our approach was to append the libraries that needed to be loaded from the header files to the shared_preload_libraries parameter in postgresql.conf.
> 
> Since the libraries loaded from the header files are compiled into the kernel, we want them to have the highest priority, If some libraries need to be loaded immediately, it could cause conflicts.
> 
> Therefore, we place the loading of libraries from the header files first, ensuring they are loaded during the compilation process before the libraries specified in the postgresql.conf configuration are loaded.

Can you give an example of what kind of problems we will encounter please? The loading order matters looks strange, let's see if there are other ways to solve it

yjhjstz commented:
How about give a hook to self adjust preload library order ?

MisterRaindrop commented:
Initially, we added the header file to load the lib because we wanted to load some specific libs. These libs remain unchanged regardless of user modifications.

If we switch to using hooks, these specific libs might fail to load, which could lead to significant issues.

MisterRaindrop commented:
> > Example:
> > Ensuring lib1 is Loaded First is Crucial
> > For instance, we have added lib1 in the header file, and in postgresql.conf, we have added lib2, lib3, and lib4.
> > Following the original approach, the loading order would be:
> > ```
> > lib2, lib3, lib4, lib1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If we require lib1 to be loaded first, this cannot be achieved because the order is lib2, lib3, lib4, lib1. In theory, the header file should have the highest loading priority.
> > After modification, the loading order will be: header file + postgresql.conf
> > ```
> > lib1, lib2, lib3, lib4
> > ```
> 
> So we require loading the modules closely related to the kernel in the Header File first, and then loading the modules in the configuration file?

Yes
